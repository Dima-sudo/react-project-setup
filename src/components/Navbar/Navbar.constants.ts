/* eslint-disable no-unused-vars */

export enum DataTestIds {
    COMPONENT_CONTAINER = 'navbar-main__container',
    USER_EMAIL = 'navbar-email__container',
    USER_AVATAR = 'navbar-avatar__image',
}
