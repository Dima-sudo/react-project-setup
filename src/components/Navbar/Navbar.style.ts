import { styled } from '@mui/material/styles';

import AppBarComponent from '@mui/material/AppBar';
import SvgIconComponent from '../SvgIcon/SvgIcon';

export const AppBar = styled(AppBarComponent)(
    ({
        theme: {
            custom: { colors },
        },
    }) => ({
        backgroundColor: colors.white,
        height: '5rem',

        '& > div': {
            height: '100%',
            '& > div': {
                height: '100%',
            },
        },
    })
);

export const UserEmail = styled('div')(
    ({
        theme: {
            custom: { colors },
        },
    }) => ({
        color: colors.text1,
        fontSize: '1rem',
        fontWeight: 500,
        paddingLeft: '1rem',
        display: 'flex',
        alignItems: 'center',
    })
);

export const UserSectionContainer = styled('div')(() => ({
    display: 'flex',
}));

export const MenuArrowIcon = styled(SvgIconComponent)<{ open: boolean }>(
    ({
        open,
        theme: {
            custom: { params },
        },
    }) => ({
        transform: open ? 'rotate(180deg)' : 'unset',
        transition: params.transitionDefault,
        paddingLeft: '0.5rem',
    })
);

export const NavbarContentContainer = styled('div')<{
    featureFlagNavbarMenu: boolean;
}>(({ featureFlagNavbarMenu }) => ({
    flexGrow: 0,
    marginLeft: 'auto',
    cursor: featureFlagNavbarMenu ? 'pointer' : 'unset',
}));

export const AvatarIcon = styled(SvgIconComponent)(() => ({
    padding: 0,
}));

export const ToolbarContainer = styled('div')(() => ({
    padding: '0 2.5rem',
}));
