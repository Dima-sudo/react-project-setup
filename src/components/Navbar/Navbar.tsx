import * as React from 'react';
import { useTranslation } from 'react-i18next';

import Toolbar from '@mui/material/Toolbar';
import Container from '@mui/material/Container';
import Tooltip from '@mui/material/Tooltip';

import NavbarMenu from './components/NavbarMenu/NavbarMenu';

import { IconNamesEnum } from '../SvgIcon/SvgIcon.constants';
import { DataTestIds } from './Navbar.constants';
import { theme } from '../../theme';
import * as S from './Navbar.style';

interface IProps {
    userEmail: string;
    menuItems?: string[];
    featureFlagNavbarMenu?: boolean;
}

const Navbar: React.FC<IProps> = ({
    userEmail,
    menuItems,
    featureFlagNavbarMenu = true,
}) => {
    const { t } = useTranslation();
    const [anchorElUser, setAnchorElUser] = React.useState<null | HTMLElement>(
        null
    );

    const handleOpenUserMenu = (event: React.MouseEvent<HTMLElement>) => {
        setAnchorElUser(event.currentTarget);
    };

    const handleCloseUserMenu = () => {
        setAnchorElUser(null);
    };

    const isMenuOpen = Boolean(anchorElUser);

    const userDetails = (
        <S.UserSectionContainer>
            <S.AvatarIcon
                data-testid={DataTestIds.USER_AVATAR}
                iconName={IconNamesEnum.AVATAR}
                onClick={
                    featureFlagNavbarMenu ? handleOpenUserMenu : () => null
                }
                height="2rem"
                width="2rem"
                isButton={featureFlagNavbarMenu}
            />

            <S.UserEmail
                onClick={handleOpenUserMenu}
                data-testid={DataTestIds.USER_EMAIL}
            >
                {userEmail}
            </S.UserEmail>

            {featureFlagNavbarMenu && (
                <S.MenuArrowIcon
                    iconName={IconNamesEnum.ARROW_DOWN}
                    open={isMenuOpen}
                    color={theme.colors.text1}
                    isButton
                    onClick={handleOpenUserMenu}
                />
            )}
        </S.UserSectionContainer>
    );

    return (
        <S.AppBar
            position="static"
            id="back-to-top-anchor"
            data-testid={DataTestIds.COMPONENT_CONTAINER}
        >
            {/* <Container maxWidth="xl" style={{ border: '1px solid red' }}> */}
            <S.ToolbarContainer>
                <Toolbar disableGutters>
                    <S.NavbarContentContainer {...{ featureFlagNavbarMenu }}>
                        {featureFlagNavbarMenu ? (
                            <Tooltip title={t('navbar.openSettingsTooltip')}>
                                {userDetails}
                            </Tooltip>
                        ) : (
                            userDetails
                        )}

                        {featureFlagNavbarMenu && (
                            <NavbarMenu
                                {...{
                                    handleCloseUserMenu,
                                    anchorElUser,
                                    isMenuOpen,
                                }}
                                items={menuItems ?? []}
                            />
                        )}
                    </S.NavbarContentContainer>
                </Toolbar>
            </S.ToolbarContainer>
            {/* </Container> */}
        </S.AppBar>
    );
};
export default Navbar;
