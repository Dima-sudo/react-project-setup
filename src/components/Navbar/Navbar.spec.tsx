import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Navbar from './Navbar';
import Setup from '../../tests/setup/setup';
import { Mocks } from './tests/mocks/Navbar.mocks';
import { DataTestIds } from './Navbar.constants';

describe('Tests Navbar component functionality', () => {
    it('Snapshots component', () => {
        render(
            <Setup>
                <Navbar userEmail={Mocks.USER_EMAIL} />
            </Setup>
        );

        expect(
            screen.getByTestId(DataTestIds.COMPONENT_CONTAINER)
        ).toMatchSnapshot();
    });

    it('Displays username correctly', () => {
        render(
            <Setup>
                <Navbar userEmail={Mocks.USER_EMAIL} />
            </Setup>
        );

        expect(
            screen.getByTestId(DataTestIds.USER_EMAIL).textContent
        ).toMatchInlineSnapshot(`"diborodin@deloitte.co.il"`);
    });

    it('Displays avatar image', () => {
        render(
            <Setup>
                <Navbar userEmail={Mocks.USER_EMAIL} />
            </Setup>
        );

        const avatar = screen.getByTestId(DataTestIds.USER_AVATAR);

        expect(avatar).toBeInTheDocument();
        expect(avatar).toMatchInlineSnapshot(`
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-uw9pdr-MuiButtonBase-root-MuiIconButton-root"
              data-testid="navbar-avatar__image"
              height="2rem"
              tabindex="0"
              type="button"
              width="2rem"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                  data-testid="PersonIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                  />
                </svg>
              </div>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
        `);
    });
});
