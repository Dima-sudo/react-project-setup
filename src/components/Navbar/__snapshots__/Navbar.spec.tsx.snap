// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Navbar component functionality Snapshots component 1`] = `
<header
  class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-1k0ixfn-MuiPaper-root-MuiAppBar-root"
  data-testid="navbar-main__container"
  id="back-to-top-anchor"
>
  <div
    class="css-v8czxn"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular css-r6ewbb-MuiToolbar-root"
    >
      <div
        class="css-qfqfuh"
      >
        <div
          aria-label="Open Settings"
          class=" css-k008qs"
          data-mui-internal-clone-element="true"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-uw9pdr-MuiButtonBase-root-MuiIconButton-root"
            data-testid="navbar-avatar__image"
            height="2rem"
            tabindex="0"
            type="button"
            width="2rem"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                data-testid="PersonIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                />
              </svg>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            class="css-1mix1ln"
            data-testid="navbar-email__container"
          >
            diborodin@deloitte.co.il
          </div>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zp4dzp-MuiButtonBase-root-MuiIconButton-root"
            fill="#575F6E"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m7 10 5 5 5-5z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</header>
`;
